lib/rules/prefer-wait-for.ts

```diff
@@ -112,46 +150,100 @@ export default ESLintUtils.RuleCreator(getDocsUrl)<Options, MessageIds>({
+        const callerIsTestingLibraryFromRequire =
+          isIdentifier(node.object) &&
+          isCallExpression(testingLibraryNode) &&
+          isVariableDeclarator(testingLibraryNode.parent) &&
+          isIdentifier(testingLibraryNode.parent.id) &&
+          node.object.name === testingLibraryNode.parent.id.name;
+        if (
+          !callerIsTestingLibraryFromImport &&
+          !callerIsTestingLibraryFromRequire
+        ) {
+          // the method does not match from the imported elements from TL (even from custom)
+          return;
+        }
+        addWaitFor = true;
+        reportWait(node.property as TSESTree.Identifier); // compiler is not picking up correctly, it should have inferred it is an identifier
       },
-      'ImportDeclaration[source.value=/testing-library/] > ImportNamespaceSpecifier'(
-        node: TSESTree.ImportNamespaceSpecifier
-      ) {
-        context.getDeclaredVariables(node).forEach((variable) =>
-          variable.references.forEach((reference) => {
-            if (
-              isMemberExpression(reference.identifier.parent) &&
-              isIdentifier(reference.identifier.parent.property) &&
-              DEPRECATED_METHODS.includes(
-                reference.identifier.parent.property.name
-              )
-            ) {
-              reportWait(reference.identifier.parent.property);
-            }
-          })
-        );
+      'CallExpression > Identifier'(node: TSESTree.Identifier) {
+        const testingLibraryNode =
+          helpers.getCustomModuleImportNode() ??
+          helpers.getTestingLibraryImportNode();
+        // this verifies the owner of the MemberExpression is the same as the node if it was imported with "import { deprecated as aliased } from 'foo'"
+        const callerIsTestingLibraryFromImport =
+          isImportDeclaration(testingLibraryNode) &&
+          testingLibraryNode.specifiers.some(
+            (s) =>
+              isImportSpecifier(s) &&
+              DEPRECATED_METHODS.includes(s.imported.name) &&
+              s.local.name === node.name
+          );
+        // this verifies the owner of the MemberExpression is the same as the node if it was imported with "const { deprecatedMethod } = require('foo')"
+        const callerIsTestingLibraryFromRequire =
+          isCallExpression(testingLibraryNode) &&
+          isVariableDeclarator(testingLibraryNode.parent) &&
+          isObjectPattern(testingLibraryNode.parent.id) &&
+          testingLibraryNode.parent.id.properties.some(
+            (p) =>
+              isProperty(p) &&
+              isIdentifier(p.key) &&
+              isIdentifier(p.value) &&
+              p.value.name === node.name &&
+              DEPRECATED_METHODS.includes(p.key.name)
+          );
+        if (
+          !callerIsTestingLibraryFromRequire &&
+          !callerIsTestingLibraryFromImport
+        ) {
+          return;
```

Comment:
Very similar to the previous one. I think this and the previous code should be merged within a detection helper.